Fix bug, add command, and RFC: drop lock.

tfogal opened the pull request:
97b84f8: simply adds a new command.  Looks like the style you wanted, clue me in if not.
d6d6d0b: the process didn't exit, so don't pretend that it did.
f356bdb: RFC: I do not think we need this lock.  If the channel is closing, that implies the process no longer exists.  If the process no longer exists, it will not send us any events.  Another way: "process is finished" is by definition the last event that a tracee can give us, and thus while we are handling "process is finished" it is not possible for the tracee to send us another event.
No?

eaburns commented:
I saw this, and I'll take a look at it in greater detail this weekend. I
think that I'm going to have feedback and that I will disagree with your
argument to ditch the locks, but I need to spend some time thinking about
it. As a teaser, here are my thoughts from what I remember when I read your
response briefly this morning:

I added the lock because Go's race detector found a race. When I looked
into it, it was an actual race, not just the detector being conservative.
Quickly looking at the code: the problem is that do() can read the commands
channel (into a register or wherever), readying to do something. Before it
gets to send the command to the channel, the Wait in the command execution
go routine can wake up claiming that the process exited (for whatever
reason---perhaps it's pessimistic to assume that this would happen without
the tracer knowing about it, but when it comes to library users, I'm
pessimistic) and close the channel. Now when the do() resumes, it will send
to a closed channel---an error!

The different ptrace commands should, indeed, return TraceeExited as an
error when the do() function returns false. The do function returns false
if the command execution goroutine exits before the do function is called.
The command execution goroutine only exits when the tracee has exited. At
worst, a naive program could just call ptrace.SingleStep (or whatever I
named the function---too lazy to check it at the second) in an infinite
loop. At some point, the tracee will exit. Subsequent calls to SingleStep
should report the error, even if it doesn't makes sense that the tracer
still bothering to call it.

On Wed May 07 2014 at 9:38:01 AM, tfogal notifications@github.com wrote:

> 97b84f8 https://github.com/eaburns/ptrace/commit/97b84f8: simply adds a
> new command. Looks like the style you wanted, clue me in if not.
> d6d6d0b https://github.com/eaburns/ptrace/commit/d6d6d0b: the process
> didn't exit, so don't pretend that it did.
> f356bdb https://github.com/eaburns/ptrace/commit/f356bdb: RFC: I do not
> think we need this lock. If the channel is closing, that implies the
> process no longer exists. If the process no longer exists, it will not send
> us any events. Another way: "process is finished" is by definition the last
> event that a tracee can give us, and thus while we are handling "process is
> finished" it is not possible for the tracee to send us another event.
> 
> ## No?
> 
> You can merge this Pull Request by running
> 
>   git pull https://github.com/tfogal/ptrace master
> 
> Or view, comment on, or merge it at:
> 
>   https://github.com/eaburns/ptrace/pull/2
> Commit Summary
> - Add 'Continue' tracee command.
> - SingleStep does not mean the process exited.
> - drop lock for synchronizing routines.
> 
> File Changes
> - _M_ ptrace.go https://github.com/eaburns/ptrace/pull/2/files#diff-0(21)
> 
> Patch Links:
> - https://github.com/eaburns/ptrace/pull/2.patch
> - https://github.com/eaburns/ptrace/pull/2.diff
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/eaburns/ptrace/pull/2
> .

eaburns commented:
Other than ditching f356bdb9 and returning TraceeExited in the command methods, it looks good.

tfogal committed:
Add a 'Continue' function.

tfogal committed:
Add a call to send a signal to the tracee.

Notably this can be used to stop the tracee (via syscall.SIGSTOP).

tfogal committed:
drop lock and add a `Close' call.

this makes the user responsible for cleanup, instead of it
happening asynchronously.

tfogal commented:
I updated this pull request.  I ended up adding the `Close` call like you mentioned.  I think I actually wasn't fully understood; I wanted both the closing of the channel and the checking for nil to both be handled in `do`.  But when I went to implement that, it turned out to be more difficult than imagined, since `do` does not (and should not) read from the channel, only write to the channel.

I fixed it at first by adding a new `finished` channel that signals that `do` should clean up.  This got the original semantics (automatic cleanup)... but I did not like it.  I implemented the `Close` and found it cleaner.

While I like the `Close`-based way better just because it's simpler, as you mentioned somewhere this doesn't seem to help with performance much.  If I run the test Go program in single-threaded mode, it is still almost 8x slower than the equivalent C program.  3x slower if I run the Go program threaded.  While I realize that Go is necessarily going to be slower, it seems like the overhead should be close to 0 in this kind of program---this kind of code should spend all its time blocked in the OS.

I need to move on today, but i'll come back to this later.

eaburns commented:
I'll try to take a look later tonight, or tomorrow, or the day after...
anyway, sometime before this weekend.

After we get this pulled in, we can think more about performance. My guess
is just that there is a bunch of waiting and extra context switching; the
Go version deals with the tracer logic in one thread and executes all of
the system calls in another. I believe that Go1.3 (coming soon, to a
computer near you) has a bunch of performance improvements that may help.
You could try the beta, which is out right now. Also Dmitry Vyukov has some
channel improvements proposed for Go1.4, I believe. They are trying to
stick to a 6mo release cycle, so maybe they'll be out before you're done
with the project. (One nice thing about a new language is that there is
still lots of room for improvement, so you can just sit back and watch your
code get faster.)

By the way, just to be clear, you said that the version with the Close
method had worse performance? Was that with respect to your C version or
with respect to the original Go version that had the race condition? If
it's the latter then I'm pretty surprised. Both the version with Close and
the racy version should be the same.

On Tue May 13 2014 at 5:48:59 AM, tfogal notifications@github.com wrote:

> I updated this pull request. I ended up adding the Close call like you
> mentioned. I think I actually wasn't fully understood; I wanted both the
> closing of the channel and the checking for nil to both be handled in do.
> But when I went to implement that, it turned out to be more difficult than
> imagined, since do does not (and should not) read from the channel, only
> write to a channel.
> 
> I fixed it at first by adding a new finished channel that signals that doshould clean up. This got the original semantics (automatic cleanup)... but
> I did not like it. I implemented the Close and found it cleaner.
> 
> While I like the Close-based way better just because it's simpler, as you
> mentioned somewhere this doesn't seem to help with performance much. If I
> run the test Go program in single-threaded mode, it is still almost 8x
> slower than the equivalent C program. 3x slower if I run the Go program
> threaded. While I realize that Go is necessarily going to be slower, it
> seems like the overhead should be close to 0 in this kind of program---this
> kind of code should spend all its time blocked in the OS.
> 
> I need to move on today, but i'll come back to this later.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/eaburns/ptrace/pull/2#issuecomment-42936233
> .

tfogal committed:
Add a "PID" method to tracee.

eaburns commented:
Phew, finally got around to looking at the code. I like it, except my few mostly-stylistic remarks. TL;DR: 2 things: 1) run ./gok.sh and fix the stuff it whines about, 2) remove Error() and have Close() return the error.

Let me know if you will never have time to make these changes. I can always pull your branch into a local one, make the changes on top of yours, and merge it in.

BTW, wow, github's interface for codereview is not very good. <--- Edit: Oops, I could've looked at the total set of changes instead of navigating to each individual commit. Not github's fault. I'll try to do it correctly next time, and maybe it'll work out better.

tfogal committed:
add way to test, and ReadWord command.

tfogal committed:
Add a 'WriteWord' call.

tfogal committed:
Add 'attach' functionality.

tfogal committed:
add missing file

should have been committed a long time ago!  oops.

tfogal committed:
add way to read instruction pointer register.

tfogal committed:
add way to set instruction pointer.

tfogal committed:
fix embarassing bug with last commit.

never actually did anything!

tfogal committed:
indent fix: we use tabs in this tree.

tfogal committed:
add a new call to grab ALL the registers.

all the things.

tfogal committed:
fix test and add hack to be SURE it sets something.

tfogal commented:
New PR created, should close this one w/o merging.
